net/graphsync_fetcher_test.go

```diff
@@ -0,0 +1,296 @@
+		builder := chain.NewBuilder(t, address.Undef)
+		gen := builder.NewGenesis()
+		final := builder.AppendOn(gen, 3)
+
+		stubs := []requestResponse{
+			requestResponse{
+				fakeRequest{cidlink.Link{Cid: final.At(0).Cid()}, layer1Selector},
+				fakeResponse{blks: []format.Node{final.At(0).ToNode()}},
+			},
+			requestResponse{
+				fakeRequest{cidlink.Link{Cid: final.At(1).Cid()}, layer1Selector},
+				fakeResponse{blks: []format.Node{final.At(1).ToNode()}},
+			},
+			requestResponse{
+				fakeRequest{cidlink.Link{Cid: final.At(2).Cid()}, layer1Selector},
+				fakeResponse{blks: []format.Node{final.At(2).ToNode()}},
+			},
+			requestResponse{fakeRequest{cidlink.Link{Cid: final.At(0).Cid()}, gsSelector}, fakeResponse{
+				responses: []graphsync.ResponseProgress{
+					makeGsResponse("", final.At(0).Cid()),
+					makeGsResponse("parents", final.At(0).Cid()),
+					makeGsResponse("parents/0", gen.At(0).Cid()),
+				},
+				blks: []format.Node{final.At(0).ToNode(), gen.At(0).ToNode()},
+			}},
+		}
+		mgs := &mockableGraphsync{stubs: stubs, t: t, store: bs}
+
+		fetcher := net.NewGraphSyncFetcher(ctx, mgs, bs, bv)
+
+		done := func(ts types.TipSet) (bool, error) {
+			if ts.Key().Equals(gen.Key()) {
+				return true, nil
+			}
+			return false, nil
+		}
+
+		ts, err := fetcher.FetchTipSets(ctx, final.Key(), peer.ID("fake"), done)
+		require.NoError(t, err, "the request completes successfully")
+		require.Equal(t, 4, len(mgs.receivedRequests), "all expected graphsync requests are made")
+		require.Equal(t, 2, len(ts), "the right number of tipsets is returned")
+		require.True(t, final.Key().Equals(ts[0].Key()), "the initial tipset is correct")
+		require.True(t, gen.Key().Equals(ts[1].Key()), "the remaining tipsets are correct")
+	})
+
+	t.Run("value returned with non block format", func(t *testing.T) {
+		notABlock := types.NewMsgs(1)[0]
+		notABlockObj, err := notABlock.ToNode()
+		require.NoError(t, err)
+
+		notABlockCid, err := notABlock.Cid()
+		require.NoError(t, err)
+		originalCids := types.NewTipSetKey(notABlockCid)
+
+		stubs := []requestResponse{
+			requestResponse{
+				fakeRequest{cidlink.Link{Cid: notABlockCid}, layer1Selector},
+				fakeResponse{blks: []format.Node{notABlockObj}},
+			},
+		}
+		mgs := &mockableGraphsync{stubs: stubs, t: t, store: bs}
+
+		fetcher := net.NewGraphSyncFetcher(ctx, mgs, bs, bv)
+
+		done := func(ts types.TipSet) (bool, error) {
+			if ts.Key().Equals(originalCids) {
+				return true, nil
+			}
+			return false, nil
+		}
+		ts, err := fetcher.FetchTipSets(ctx, originalCids, peer.ID("fake"), done)
+		require.Errorf(t, err, "fetched data (cid %s) was not a block", notABlockCid.String())
+		require.Nil(t, ts)
+	})
+}
+
+func TestRealWorldGraphsyncFetchAcrossNetwork(t *testing.T) {
+	tf.IntegrationTest(t)
+	ctx := context.Background()
+	// setup a chain
+	builder := chain.NewBuilder(t, address.Undef)
+	gen := builder.NewGenesis()
+	final := builder.AppendManyOn(30, gen)
```

Comment:
nit: Can the `30` be put into a variable and then make `31` ([here](https://github.com/filecoin-project/go-filecoin/pull/3153/files#diff-883ff976e50c0f0232bc503f90d13111R220)) `variable+1`?